samples/bluetooth/iso_broadcast/src/main.c

```diff
@@ -197,23 +206,117 @@ int main(void)
+				ret = bt_iso_chan_send_ts(&bis_iso_chan[chan], buf, seq_num[chan],
+							  ts);
+			} else {
+				seq_num[chan] = 0U;
+				ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, 0U);
+			}
+
 			if (ret < 0) {
 				printk("Unable to broadcast data on channel %u"
 				       " : %d", chan, ret);
 				net_buf_unref(buf);
 				return 0;
 			}
 
+			/* Get Tx Sync, to verify that all streams have their ISO data at the same
+			 * timestamp.
+			 */
+			ret = bt_iso_chan_get_tx_sync(&bis_iso_chan[chan], &tx_info);
+			if (ret) {
+				printk("Failed to get tx sync after send (%d)\n", ret);
+				return 0;
+			}
+
+			/* Timestamp delta with the Controller */
+			if (!delta) {
+				delta = tx_info.ts - ts;
+
+			/* Use one channel timestamp to accumulate relative drift */
+			} else if (chan == 0U) {
+				/* NOTE: Change in delta (drift) for a duration of SDU intervals can
+				 *       be used to synchronize the clock with the Controller.
+				 */
+				static uint8_t countdown = SOURCE_DRIFT_COMP_COUNTDOWN;
+				static int64_t drift_us;
+
+				/* Accumulate relative drift */
+				drift_us += delta - tx_info.ts + ts;
+
+				/* Transfer the accumulated drift to compensate sampling interval */
+				if (!countdown--) {
+					countdown = SOURCE_DRIFT_COMP_COUNTDOWN;
+					drift_comp_us = drift_us;
+					drift_us = 0U;
+
+					printk("Compensate %lld us drift\n", drift_comp_us);
+				}
+			}
+
+			/* Permit a reasonable clock jitter (at least Bluetooth active clock
+			 * jitter), in the verification of timestamp.
+			 *
+			 * Lets consider that the source clock is allowed to drift up to the
+			 * BIG SDU interval before drift compensation is applied.
+			 */
+			uint32_t jitter_us = BIG_SDU_INTERVAL_US;
+
+			/* Validate ISO data being enqueued at the required timestamp.
+			 * Ignore checking first ISO data sent.
+			 */
+			if (!is_first_iso_data_send &&
+			    !IN_RANGE(tx_info.ts, (ts + delta - jitter_us),
+				      (ts + delta + jitter_us))) {
+				printk("Mismatch in tx sync timestamp! (%llu != %u)\n",
+				       (ts + delta), tx_info.ts);
+				return 0;
+			}
+
+			seq_num[chan]++;
 		}
 
+		is_first_iso_data_send = false;
+
 		if ((iso_send_count % CONFIG_ISO_PRINT_INTERVAL) == 0) {
 			printk("Sending value %u\n", iso_send_count);
 		}
 
 		iso_send_count++;
-		seq_num++;
 
+
+		/* NOTE: Here we are simulating a drift compensated sleep for
+		 *       BIG_SDU_INTERVAL_US
+		 */
+		uint32_t sleep_us;
+
+		if (drift_comp_us < BIG_SDU_INTERVAL_US) {
+			sleep_us = BIG_SDU_INTERVAL_US - drift_comp_us;
+		} else {
+			sleep_us = 0U;
+		}
+
+		/* Reset drift compensation value */
+		drift_comp_us = 0;
+
+		/* Sleep, simulating source ISO data sampling */
+		uint32_t ts_curr;
+
+		do {
+			k_sleep(K_USEC(10));
+			ts_curr = k_ticks_to_us_floor32(sys_clock_tick_get());
+		} while ((ts_curr - ts_prev) < sleep_us);
+
+		/* Increment the application timestamp by the simulated sampling duration */
+		ts += (ts_curr - ts_prev);
+		ts_prev = ts_curr;
```

Comment:
Ditto here: What is the purpose here? 

What's the reason to have a sleep here?

Reply:
The sample provides examples for the 4 statements in the Bluetooth Core Specification v6.0, Vol 6, Part G, Section 3.3 Time Stamp for SDU:

1. `should`: Time of arrival (use of `is_first_iso_data_send` and subsequent use of simulated `k_sleep()`)
```
When an HCI ISO Data packet sent by the Host does not contain a Time_Stamp or
the Time_Stamp value is not based on the Controller's clock, the Controller should
determine the CIS or BIS event to be used to transmit the SDU contained in that packet
based on the time of arrival of that packet.
``` 
2. `shall`: Time_Stamp corresponding to the Packet_Sequence_Number (verification of `ts` to match `tx_info.ts` for every `seq_num`)
```
In each direction, in a given packet, the Time_Stamp provided (if any) shall be one that
applies to the SDU corresponding to the Packet_Sequence_Number.
```
3. `should`: Compensate potential clock drift and jitter (calculation of `drift_comp_us`)
```
When using the Time_Stamp or another suitable method for synchronization, both the
higher layer and Controller should implement a mechanism to compensate potential
clock drift and jitter. These mechanisms should include long term drift compensation
and tolerances on the synchronization method used.
```
4. `may`: Use of timestamps at the Controller’s clock
```
The Controller or the higher layer sending or receiving an SDU may use any of the
following methods to transfer the synchronization reference of the SDU to the higher
layer:
• Timestamps at the Controller’s clock
• Any other signaling method suitable to the higher layer
```